packages/element/src/create-interpolate-element.js

```diff
@@ -0,0 +1,326 @@
+ *
+ * match: Full match of the tag (i.e. <strong>, </strong>, <br/>)
+ * isClosing: If the match has a closing tag (i.e. </)
+ * name: The name portion of the tag (strong, br) (if )
+ * isSelfClosed: Whether the match is for a self-closing tag (i.e. />)
+ *
+ * @type RegExp
+ */
+const tokenizer = /<(\/)?(\w+)\s*(\/)?>/g;
+
+/**
+ * This receives the value for a map element which consists of an element and
+ * it's props and returns a element creator function.
+ *
+ * @param {Array} [ element, props ] The first item in the array is expected to
+ *                                   be a string or a react component. The
+ *                                   second item is expected to be either
+ *                                   undefined or an object.
+ */
+const elementCreator = ( [ element, props = {} ] ) => ( children ) => {
+	if ( typeof element !== 'string' && typeof element !== 'function' ) {
+		throw new Error( `Not a valid element (${ element })` );
+	}
+	props.key = ++keyIndex;
+	return createElement( element, props, children );
+};
+
+/**
+ * An object describing a component to be created.
+ *
+ * This is used by the string iterator to track children that get added to an
+ * element when it is created. This allows for collecting nested elements in
+ * the string before creating the parent.
+ *
+ * @param   {Function}  creator An element creator that will be invoked for
+ *                              actually creating the react element with the
+ *                              provided children.
+ * @param   {Array}  children   Array of children to be provided to the element
+ *                              creator when invoked.
+ *
+ * @return  {Object}            An object returning the creator and children.
+ */
+function Component( creator, children = [] ) {
+	return {
+		creator,
+		children,
+	};
+}
+
+/**
+ * This encapsulates information about the current iteration state for adding to
+ * the stack
+ *
+ * @param {Component} component
+ * @param {number}    tokenStart
+ * @param {number}    tokenLength
+ * @param {number}    prevOffset
+ * @param {number}    leadingTextStart
+ *
+ * @return {Object} Iteration info as an object.
+ */
+function Frame(
+	component,
+	tokenStart,
+	tokenLength,
+	prevOffset,
+	leadingTextStart
+) {
+	return {
+		component,
+		tokenStart,
+		tokenLength,
+		prevOffset: prevOffset || tokenStart + tokenLength,
+		leadingTextStart,
+	};
+}
+
+/**
+ * This function creates an interpolated element from a passed in string with
+ * specific tags matching how the string should be converted to an element via
+ * the conversion map value.
+ *
+ * @example
+ * For example, for the given string:
+ *
+ * "This is a <span>string</span> with <a>a link</a> and a self-closing
+ * <CustomComponentB/> tag"
+ *
+ * You would have something like this as the conversionMap value:
+ *
+ * ```js
+ * {
+ *     span: ['span'],
+ *     a: ['a', { href: 'https://github.com' }],
+ *     CustomComponentB: [ CustomComponentB ],
```

Comment:
If we pass-in JSX here, which would be equivalent to `createElement( 'span' )` then the interface would probably be more natural.

in person at WCUS it wasn't obvious to me how to incorporate the children into the elements but we should be able to do this with `React.cloneElement()`, or whatever internal Gutenberg has wrapped that with.

```js
function Component( element, children ) {
	return { element, children };
}
```

```js
doItToIt(
	'This is a <span>string</span> with <a>a link</a> and a self-closing <CustomComponentB/> tag',
	{
		span: <span style={{ color: 'red' }} />,
		a: <a href="https://github.com" />,
		CustomComponentB: <CustomComponentB id={ 14 } />,
	}
)
```

then…

```js
output.push( React.cloneElement( component.element, null, component.children ) );
```

Reply:
(See also my comment https://github.com/WordPress/gutenberg/pull/17376#issuecomment-550320856)

The Gutenberg docs are both ES5 and ESNext, the former advocates for a `wp.element.createElement()` syntax which seems to be equivalent of React's `wp.element.createElement ()`. If we follow the documentation on this, we'd just go for accepting `wp.element.createElement()` and JSX.

Reply:
I like this idea.  Performance wise using `wp.element.cloneElement` shouldn't be much of a hit.  It also allows for using `wp.element.isValidElement` to typecheck the map value.

Reply:
Ya this idea really helps simplify the syntax for the arguments (and with the bonus of more strict map validation...linting rules will be easier to right too I imagine).  See 71fb940